docs/serialisation/protobuf.md

```diff
@@ -0,0 +1,410 @@
+
+Smithy:
+
+```smithy
+structure Foo {
+  @required
+  uuid : alloy#UUID
+}
+
+```
+
+Proto:
+```proto
+message UUID {
+  int64 upper_bits = 1;
+  int64 lower_bits = 2;
+}
+
+message Foo {
+  uuid: UUID
+}
+```
+
+#### Aggregate Types
+
+##### Structure
+
+Smithy:
+```smithy
+structure Testing {
+  myString: String,
+  myInt: Integer
+}
+```
+
+Proto:
+```proto
+import "google/protobuf/wrappers.proto";
+
+message Testing {
+  google.protobuf.StringValue myString = 1;
+  google.protobuf.Int32Value myInt = 2;
+}
+```
+
+##### Union
+
+Unions in Smithy are tricky to translate to Protobuf because of the nature of `oneOf` : unions are first-class citizens in Smithy, whereas `oneOf` can only exist relatively to messages in proto. Therefore, the default encoding for unions in protobuf is equivalent to the one of a proto `message` that contains a `definition` field which is the `oneOf`. For example:
+
+Smithy:
+```smithy
+structure Union {
+  @required
+  value: TestUnion
+}
+
+union TestUnion {
+    num: Integer,
+    txt: String
+}
+```
+
+Proto:
+```proto
+message Union {
+  foo.TestUnion value = 1;
+}
+
+message TestUnion {
+  oneof definition {
+    int32 num = 1;
+    string txt = 2;
+  }
+}
+```
+
+It is possible to use the `alloy@protoInlinedOneOf` to indicate that a union should be encoded as if the corresponding `oneOf` was directly inlined in a message. This is subject to additional constraints, are `oneOf` field indices are supposed to flattened into the containing `message`'s' field indices. However, this encoding is more compact.
+
+For example:
+
+Smithy:
+```smithy
+
+use alloy.proto#protoInlinedOneOf
+
+structure Union {
+  @required
+  value: TestUnion
+}
+
+@protoInlinedOneOf
+union TestUnion {
+    num: Integer,
+    txt: String
+}
+```
+
+Proto:
+```proto
+syntax = "proto3";
+
+package foo;
+
+message Union {
+  oneof value {
+    int32 num = 1;
+    string txt = 2;
+  }
+}
+```
+
+##### List
+
+Smithy:
+```smithy
+list StringArrayType {
+    member: String
+}
+structure StringArray {
+    value: StringArrayType
+}
+```
+
+Proto:
+```proto
+message StringArray {
+  repeated string value = 1;
+}
+```
+
+##### Map
+
+Smithy:
+```smithy
+map StringStringMapType {
+    key: String,
+    value: String
+}
+structure StringStringMap {
+  value: StringStringMapType
+}
+```
+
+Proto:
+```proto
+message StringStringMap {
+  map<string, string> value = 1;
+}
+```
+
+##### String Enum (closed)
+
+Smithy:
+```smithy
+enum Color {
+    RED
+    GREEN
+    BLUE
+}
+```
+
+Proto:
+```proto
+enum Color {
+  RED = 0;
+  GREEN = 1;
+  BLUE = 2;
+}
+```
+
+##### String Enum (open)
+
+Open string enumerations are considered as raw strings when serialised to protobuf :
+
+Smithy:
+```smithy
+@alloy#openEnum
+enum Color {
+    RED
+    GREEN
+    BLUE
+}
+
+structure Foo {
+    color: Color
+}
+```
+
+Proto:
+```proto
+message Foo {
+  String color = 1;
+}
+```
+
+##### Int Enum (closed)
```

Comment:
for enum doc, I'd probably add the following:

1. for the closed enum => if you had a protoIndex to one member of the enum, you'll need to define it for all members. you'll also need to have one for the `0` value
2. for the open enum => you can't define protoIndex for open enum as we use a primitive (string / int) value to represent those

Reply:
This is talked about it a later section dedicated to [the use of `protoIndex` in the context of enumerations](https://github.com/disneystreaming/alloy/pull/129/files#diff-a0c6700b512dfe40b2006b14a53956edd38e2e6e13a34aee16e1ac4a889b2b80R341)

Reply:
ahh sorry completely overlooked that section

I was like: well he did not write that whole protobuf section, it was moved, but he must have edited the enum section

I did not see you added one 👍